scrapy/exporters.py

```diff
@@ -247,6 +247,12 @@ def export_item(self, item):
         values = list(self._build_row(x for _, x in fields))
         self.csv_writer.writerow(values)
 
+    def finish_exporting(self):
+        # Detaching stream in order to avoid file closing.
+        # The file will be closed with slot.storage.store
+        # https://github.com/scrapy/scrapy/issues/5043
+        self.stream.detach()
```

Comment:
Does this need to happen on finish exporting, or can it happen on `__init__`, right after initializing `self.stream`?

Reply:
> Does this need to happen on finish exporting, or can it happen on `__init__`, right after initializing `self.stream`?

I haven't actually tried that, but I doubt that it will work. From the [documentation](https://docs.python.org/3/library/io.html):
**_After the raw stream has been detached, the buffer is in an unusable state._**
and thus  it seems to me that it would make sense to detach the stream with `finish_exporting`.

> Maybe it is a race condition, that in the tests we are closing `file` before the exporter class gets destroyed?

I am pretty certain that this is the case since while testing I managed to avoid the problem without detaching input stream by sleeping feedexport for a few seconds after slot is closed, but before a new batch is started:
   https://github.com/scrapy/scrapy/blob/c56caaf0b76939729fdc6d8982f4b249edb7d830/scrapy/extensions/feedexport.py#L437-L443
   
 We might need to sleep the code somewhere in order to obtain successful tests?